File: backend/app/api/v1/endpoints/ai_analyst.py
"""AI analyst endpoints.

Routes stay thin: authorization runs through the shared dependency and
``dataset_access``; the pipeline lives in ``ai_analyst_service``. Deterministic
analysis is always returned even when the AI provider is unavailable.
"""

from __future__ import annotations

import uuid

from fastapi import APIRouter

from app.api.deps import CurrentUser, DbSession, Storage
from app.core.rate_limit import AiRateLimit
from app.schemas.ai_analyst import (
    AnalystAnswerResponse,
    AnalystQuestionRequest,
    AnalystReport,
    AnalyzeRequest,
)
from app.schemas.common import ErrorResponse
from app.services import ai_analyst_service

router = APIRouter(
    prefix="/projects/{project_id}/datasets/{dataset_id}/ai-analyst",
    tags=["ai-analyst"],
)

_RESPONSES: dict[int | str, dict[str, object]] = {
    401: {"model": ErrorResponse, "description": "Missing or invalid token"},
    404: {"model": ErrorResponse, "description": "Project, dataset or version not accessible"},
    422: {"model": ErrorResponse, "description": "Dataset not ready, empty, or unanalysable"},
}


@router.post(
    "/analyze",
    response_model=AnalystReport,
    summary="Analyse a dataset and generate insights",
    dependencies=[AiRateLimit],
    responses=_RESPONSES,
)
async def analyze(
    project_id: uuid.UUID,
    dataset_id: uuid.UUID,
    payload: AnalyzeRequest,
    session: DbSession,
    current_user: CurrentUser,
    storage: Storage,
) -> AnalystReport:
    """Deterministic insights first, then optional AI interpretation.

    If the AI provider is missing, misconfigured, slow or failing, the report
    is still returned with `ai_available: false` and a reason in `ai_status`.
    """
    return await ai_analyst_service.analyze(
        session, storage, current_user, project_id, dataset_id, payload
    )


@router.post(
    "/ask",
    response_model=AnalystAnswerResponse,
    summary="Ask a follow-up question about the analysis",
    dependencies=[AiRateLimit],
    responses=_RESPONSES,
)
async def ask(
    project_id: uuid.UUID,
    dataset_id: uuid.UUID,
    payload: AnalystQuestionRequest,
    session: DbSession,
    current_user: CurrentUser,
    storage: Storage,
) -> AnalystAnswerResponse:
    """Answers from the same analytical context the report was built from.

    This is not a natural-language query engine: no query is generated or run
    against the data.
    """
    return await ai_analyst_service.answer_question(
        session,
        storage,
        current_user,
        project_id,
        dataset_id,
        payload.version_id,
        payload.question,
    )

File: frontend/src/api/aiAnalyst.ts
/** AI analyst endpoints. */

import { apiClient } from '@/lib/apiClient';
import type { AnalystAnswer, AnalystReport } from '@/types/api';

function basePath(projectId: string, datasetId: string): string {
  return `/projects/${projectId}/datasets/${datasetId}/ai-analyst`;
}

export function analyzeDataset(
  projectId: string,
  datasetId: string,
  options: { versionId?: string; includeAi?: boolean; refresh?: boolean } = {},
): Promise<AnalystReport> {
  return apiClient.post<AnalystReport>(
    `${basePath(projectId, datasetId)}/analyze`,
    {
      version_id: options.versionId ?? null,
      include_ai: options.includeAi ?? true,
      refresh: options.refresh ?? false,
    },
    // Analysis runs the full pipeline, so allow more than the default timeout.
    { timeoutMs: 120_000 },
  );
}

export function askAnalyst(
  projectId: string,
  datasetId: string,
  question: string,
  versionId?: string,
): Promise<AnalystAnswer> {
  return apiClient.post<AnalystAnswer>(
    `${basePath(projectId, datasetId)}/ask`,
    { question, version_id: versionId ?? null },
    { timeoutMs: 120_000 },
  );
}
